src/DefaultBuilder/src/WebHost.cs

```diff
@@ -223,14 +223,19 @@ internal static void ConfigureWebDefaults(IWebHostBuilder builder)
             }
         });
 
-        UseKestrel(builder);
+        ConfigureWebDefaultsCore(builder);
+
+        builder.ConfigureServices(services =>
+        {
+            services.AddRouting();
```

Comment:
Can we add `RegexInlineRouteConstraintSetup` directly since we just called `ConfigureWebDefaultsCore/AddRoutingCore`? That way we don't resscan the `IServiceCollection` for a *bunch* of types. Maybe make it shared source and leave a note `AddRouting`, that any changes there need to be duplicated in WebHost.cs.

Reply:
Unfortunately we can't add `RegexInlineRouteConstraintSetup` directly because it is internal in a different assembly. I had a look at doing shared source like you mentioned but there are a few things blocking that:

1. To pull across `RegexInlineRouteConstraintSetup` we also need to pull across `RegexErrorStubRouteConstraint`.
2. `RegexErrorStubRouteConstraint` has a dependency on the resource strings
3. Doing this also breaks out ability to remove the `RegexErrorStubRouteConstraint` if someone does call `AddRoutng` to add Regex back in (which is one of the things we wanted to make possible). It breaks because the type check we do on the constraint to determine whether the user has replaced it with their own doesn't work.
4. `RegexInlineRouteConstraintSetup` also needs to access `TrimmerSafeConstraintMap` on `RouteOptions`.

I think this is a case of the cure is worse than the disease.

Reply:
... do we have precedent for AddXYZ methods actually checking to see whether they've been called already and bailing early?

Reply:
@davidfowl what are you thoughts on this? This is a lot of complexity just to overcome a test case for MVC which deliberately called `AddRouting` multiple times.